src/main_interface.jl

```diff
@@ -30,33 +22,509 @@ end
+It is best to pass an explicit version, since `HashVersion{4}` is the only non-deprecated
+version; it is much faster than 1 and more stable than 2. Furthermore, a new hash version is
+provided in a future release, the hash you get by passing an explicit `HashVersion{N}`
+should *not* change. (Note that the number in `HashVersion` does not necessarily match the
+package version of `StableHashTraits`).
 
-Instead of passing a context, you can instead pass a `version` keyword, that will set the
-context to `HashVersion{version}()`.
+In hash version 4, you customize how hashes are computed using [`transformer`](@ref), and in
+versions 1-4 using [`hash_method`](@ref).
 
 To change the hash algorithm used, pass a different function to `alg`. It accepts any `sha`
 related function from `SHA` or any function of the form `hash(x::AbstractArray{UInt8},
 [old_hash])`.
 
-The `context` value gets passed as the second argument to [`hash_method`](@ref), and as the
-third argument to [`StableHashTraits.write`](@ref)
+The `context` value gets passed as the second argument to [`hash_method`](@ref) and
+[`transformer`](@ref), and as the third argument to [`StableHashTraits.write`](@ref). Note
+that both `hash_method` and `StableHashTraits.write` are deprecated.
+
+## See Also
 
-!!! danger
-    Hash version 1 and 2 have been found to depend on the module in which they are first
-    executed (cf [#64](https://github.com/beacon-biosignals/StableHashTraits.jl/pull/64));
-    to avoid hash instability, use version 3.
+[`transformer`](@ref)
 """
 stable_hash(x; alg=sha256, version=1) = return stable_hash(x, HashVersion{version}(); alg)
 function stable_hash(x, context; alg=sha256)
-    return compute_hash!(stable_hash_helper(x, HashState(alg, context), context,
-                                            hash_method(x, context)))
+    if root_version(context) < 4
+        return compute_hash!(deprecated_hash_helper(x, HashState(alg, context), context,
+                                                    hash_method(x, context)))
+    else
+        hash_state = hash_type_and_value(x, HashState(alg, context), context)
+        return compute_hash!(hash_state)
+    end
+end
+
+"""
+    StableHashTraits.parent_context(context)
+
+Return the parent context of the given context object. (See [`hash_method`](@ref) and
+[`@context`](@ref) for details of using context). The default method falls back to returning
+`HashVersion{1}`, but this is flagged as a deprecation warning; in the future it is expected
+that all contexts define this method.
+
+This is normally all that you need to know to implement a new context. However, if your
+context is expected to be the root context—one that does not fallback to any parent (akin to
+`HashVersion`)—then there may be a bit more work involved. In this case, `parent_context`
+should return `nothing`. You will also need to define
+[`StableHashTraits.root_version`](@ref).
+"""
+function parent_context(x::Any)
```

Comment:
from pair review- after breaking release, we will delete this method or have it throw